# Variables

### Variables are fundamental elements in programming used to store data that can be referenced and manipulated in a program. In Python, variables are created when you assign a value to them, and they do not need explicit declaration to reserve memory space. The declaration happens automatically when you assign a value to a variable.

In [ ]:
## Declaring And Assigning Variables

age=26
height=6.1
name="Sai Sampath"
is_student=False

## printing the variables

print("age :",age)
print("Height:",height)
print("Name:",name)

age : 26
Height: 6.1
Name: {name} Sai Sampath


In [10]:
## Naming Conventions
## Variable names should be descriptive.
## They must start with a letter or an '_' and contains letter,numbers and underscores
## variables names case sensitive

#valid variable names

first_name="Sai"
last_name="Sampath"
#print(First_name) # will give error as variable names are case sensitive
print("First Name:",first_name)
print("Last Name:",last_name)
print("Full Name:",first_name + " " + last_name)

First Name: Sai
Last Name: Sampath
Full Name: Sai Sampath


In [ ]:
# Invalid variable names
#2age=30 #starts with number
#first-name="Sai" #contains hyphen
#is student=True #contains space
#@name="Sai" #contains special character

Cell In[6], line 4
    @name="Krish"
     ^
SyntaxError: invalid syntax. Maybe you meant '==' or ':=' instead of '='?

In [ ]:
name = "Sai"
Name = "Sampath" # Different variable due to case sensitivity

In [ ]:
# Variable Types
#Python is Dynamically Typed Language, which means you don't need to declare the variable type explicitly. The interpreter infers the type based on the assigned value.
age = 30           # Integer
height = 6.1      # Float
name = "Sai"    # String
is_student = True # Boolean
# Python has no separate double, but its float is already double precision by default.

""" 
| Concept             | C# Type   | Precision     | Python Equivalent    |
| ------------------- | --------- | ------------- | -------------------  |
| Single precision    | `float`   | ~7 digits     | ❌ Not available     |
| Double precision    | `double`  | ~15–16 digits | ✅ `float`           |
| Arbitrary precision | `decimal` | Configurable  | ✅ `decimal.Decimal` | 
"""
# Create a floating-point number
pi = 3.141592653589793
# Check the precision and size
print(f"Exact precision: {pi:.20f}")



Exact precision: 3.14159265358979311600


In [15]:
# Python float precision demo
a = 1.1234567890123456789
print(a)

# To check how many digits are preserved
print(f"Exact precision: {a:.20f}")

1.1234567890123457
Exact precision: 1.12345678901234569125


# 🧮 Why Python Shows Extra Digits in Float Values

### ❓ Question
Why does Python show extra digits like  
`1.1234567890123457` or `1.12345678901234569125`  
when I enter `1.1234567890123456789`?

---

### 💡 Explanation

- Python’s `float` is a **64-bit IEEE-754 double-precision** number.  
- It stores values in **binary**, not decimal.  
- Most decimal fractions **cannot be represented exactly** in binary.  
- Therefore, Python stores the **closest possible binary approximation**.  
- When converting back to decimal for display, you see **tiny rounding differences** — those are not extra precision, just conversion artifacts.

---

### 🧩 Example
```python
a = 1.1234567890123456789
print(a)               # 1.1234567890123457
print(f"{a:.20f}")     # 1.12345678901234569125


In [20]:
# Type check and Conversion
age = 26
print(f"Age :",age ,"Type of age:", {type(age)})  # Output: Age : 26 Type of age: {<class 'int'>}
age_str = str(age)
print(f"Converted age to string: {age_str}, Type: {type(age_str)}")  
# Output: Converted age to string: 26, Type: <class 'str'>

Age : 26 Type of age: {<class 'int'>}
Converted age to string: 26, Type: <class 'str'>


In [22]:
# String to Int
age = '26'
print(f"Age :",age ,"Type of age:", {type(age)})  # Output: Age : 26 Type of age: {<class 'str'>}
age = int(age) 
print(f"Age :",age ,"Type of age:", {type(age)}) # Output: Age : 26 Type of age: {<class 'int'>}

Age : 26 Type of age: {<class 'str'>}
Age : 26 Type of age: {<class 'int'>}


In [ ]:
name = 'Sai'
# name = int(name)  # This will raise a ValueError : invalid literal for int() with base 10: 'Sai'


ValueError: invalid literal for int() with base 10: 'Sai'

In [2]:
height = 6.1
height = int(height)  # This will convert float to int by truncating the decimal part
print(f"Height :",height ,"Type of height:", {type(height)}) # Output: Height : 6 Type of height: {<class 'int'>}

height = float(height)  # Converting back to float
print(f"Height :",height ,"Type of height:", {type(height)}) # Output: Height : 6.0 Type of height: {<class 'float'>}

Height : 6 Type of height: {<class 'int'>}
Height : 6.0 Type of height: {<class 'float'>}


In [ ]:
#Dynamic typing
#python allows you to change the type of a variable by simply assigning a new value of a different type to it.
var = 10          # Initially an integer
print(f"Value: {var}, Type: {type(var)}")  # Output: Value: 10, Type: <class 'int'>
var = "Hello"    # Now a string
print(f"Value: {var}, Type: {type(var)}")  # Output: Value: Hello, Type: <class 'str'>
var = 3.14       # Now a float
print(f"Value: {var}, Type: {type(var)}")  # Output: Value: 3.14, Type: <class 'float'>

Value: 10, Type: <class 'int'>
Value: Hello, Type: <class 'str'>
Value: 3.14, Type: <class 'float'>


In [ ]:
#input from user
age = input("Enter your age: ")  # input() returns a string
print(f"Your age is: {age}, Type: {type(age)}")  # Output: Your age is: <entered age>, Type: <class 'str'>
# To convert the input to an integer
age = int(age)
print(f"Your age is: {age}, Type: {type(age)}")  # Output: Your age is: <entered age>, Type: <class 'int'>

Your age is: 33, Type: <class 'str'>
Your age is: 33, Type: <class 'int'>


In [ ]:
### Simple Calculator
num1 = float(input("Enter first number: "))
num2 = float(input("Enter second number: "))
sum_result = num1 + num2
print(f"The sum of {num1} and {num2} is: {sum_result}")
diff_result = num1 - num2
print(f"The difference of {num1} and {num2} is: {diff_result}")
prod_result = num1 * num2
print(f"The product of {num1} and {num2} is: {prod_result}")
div_result = num1 / num2
print(f"The division of {num1} by {num2} is: {div_result}")

The sum of 10.0 and 2.0 is: 12.0
The difference of 10.0 and 2.0 is: 8.0
The product of 10.0 and 2.0 is: 20.0
The division of 10.0 by 2.0 is: 5.0
